pkg/k8s/config.go

```diff
@@ -30,13 +38,64 @@ type configuration struct {
 
 	// Burst is the burst to pass to the kubernetes client configuration.
 	Burst int
+
+	lock.RWMutex
 }
 
-// GetAPIServerURL returns the configured API server URL address
-func GetAPIServerURL() string {
+// GetAPIServerURL returns the configured API server URL address.
+func GetAPIServerURL() *url.URL {
+	config.RLock()
+	defer config.RUnlock()
+
 	return config.APIServerURL
 }
 
+// GetAPIServerURLString returns the string representation of configured API server
+// address.
+func GetAPIServerURLString() string {
+	config.RLock()
+	defer config.RUnlock()
+
+	if config.APIServerURL == nil {
+		return ""
+	}
+	return config.APIServerURL.String()
+}
+
+// RotateAPIServerURL rotates the kubernetes API server URL used for clients.
+// This is used when the currently registered kubernetes API server instance
+// is not responding(failing heartbeat check).
+func RotateAPIServerURL() {
+	config.Lock()
+	defer config.Unlock()
+
+	if len(config.APIServerURLs) < 2 {
+		return
+	}
+
+	log.WithField("oldAddress", config.APIServerURL.String()).Info("Rotating API server address")
+	rand.Shuffle(len(config.APIServerURLs), func(i, j int) {
+		config.APIServerURLs[i], config.APIServerURLs[j] = config.APIServerURLs[j], config.APIServerURLs[i]
+	})
+
+	for _, urlStr := range config.APIServerURLs {
+		if urlStr != config.APIServerURL.String() {
+			apiServerUrl, err := url.ParseRequestURI(urlStr)
+			if err != nil {
+				// Logging this as an error as this should never happen.
+				// We are already parsing APIServer URLs during configuration phase.
+				// All strings in config.APIServerURLs should be valid URLs
+				log.WithError(err).Error("Error parsing APIServer url")
```

Comment:
It's an `Error()` already
```suggestion
				log.WithError(err).Error("Failed to parse APIServer URL")
```